diff -r c7bf150bfa46 src/vm/string.cpp
--- a/src/vm/string.cpp	Fri Mar 11 23:35:56 2011 +0100
+++ b/src/vm/string.cpp	Fri Apr 01 08:23:42 2011 +0200
@@ -505,7 +505,7 @@
 			/* string already in hashtable, free memory */
 
 			if (!copymode)
-				mem_free(a, sizeof(java_chararray_t) + sizeof(u2) * (length - 1) + 10);
+				mem_free(a, sizeof(java_chararray_t) + sizeof(u2) * (length - 1) );
 
 			mutex->unlock();
 
@@ -520,11 +520,11 @@
 	java_chararray_t* acopy;
 	if (copymode) {
 		/* create copy of u2-array for new javastring */
-		u4 arraysize = sizeof(java_chararray_t) + sizeof(u2) * (length - 1) + 10;
+		u4 arraysize = sizeof(java_chararray_t) + sizeof(u2) * (length - 1);
 		acopy = (java_chararray_t*) mem_alloc(arraysize);
 /*    		memcpy(ca, a, arraysize); */
 		memcpy(&(acopy->header), &(((java_chararray_t*) a)->header), sizeof(java_array_t));
-		memcpy(&(acopy->data), &(((java_chararray_t*) a)->data) + offset, sizeof(u2) * (length - 1) + 10);
+		memcpy(&(acopy->data), &(((java_chararray_t*) a)->data) + offset, sizeof(u2) * (length - 1));
 	}
 	else {
 		acopy = (java_chararray_t*) a;
@@ -638,7 +638,7 @@
 	utflength = utf_get_number_of_u2s(u);
 
 	/* allocate memory */ 
-	a = (java_chararray_t*) mem_alloc(sizeof(java_chararray_t) + sizeof(u2) * (utflength - 1) + 10);
+	a = (java_chararray_t*) mem_alloc(sizeof(java_chararray_t) + sizeof(u2) * (utflength - 1));
 
 	/* convert utf-string to u2-array */
 	for (i = 0; i < utflength; i++)
